commit 044dabfd92d09de4f168a36a07ac3232f5647a1d
Author: Jeremy Allison <jra@samba.org>
Date:   Fri May 1 12:50:51 2015 -0700

    s3: smbd: VFS: Add vfs_stat_smb_basename() - to be called when we *know* stream name parsing has already been done.
    
    BUG: https://bugzilla.samba.org/show_bug.cgi?id=11249
    
    Signed-off-by: Jeremy Allison <jra@samba.org>
    Reviewed-by: Andreas Schneider <asn@samba.org>
    Reviewed-by: Ralph Boehme <slow@samba.org>

diff --git a/source3/smbd/proto.h b/source3/smbd/proto.h
index a5144d5..5a7d16d 100644
--- a/source3/smbd/proto.h
+++ b/source3/smbd/proto.h
@@ -1179,6 +1179,8 @@ int vfs_stat_smb_fname(struct connection_struct *conn, const char *fname,
 		       SMB_STRUCT_STAT *psbuf);
 int vfs_lstat_smb_fname(struct connection_struct *conn, const char *fname,
 			SMB_STRUCT_STAT *psbuf);
+int vfs_stat_smb_basename(struct connection_struct *conn, const char *fname,
+			SMB_STRUCT_STAT *psbuf);
 NTSTATUS vfs_stat_fsp(files_struct *fsp);
 NTSTATUS vfs_chown_fsp(files_struct *fsp, uid_t uid, gid_t gid);
 NTSTATUS vfs_streaminfo(connection_struct *conn,
diff --git a/source3/smbd/vfs.c b/source3/smbd/vfs.c
index ebd3440..b852550 100644
--- a/source3/smbd/vfs.c
+++ b/source3/smbd/vfs.c
@@ -1331,6 +1331,32 @@ int vfs_lstat_smb_fname(struct connection_struct *conn, const char *fname,
 }
 
 /**
+ * XXX: This is temporary and there should be no callers of this once
+ * smb_filename is plumbed through all path based operations.
+ *
+ * Called when we know stream name parsing has already been done.
+ */
+int vfs_stat_smb_basename(struct connection_struct *conn, const char *fname,
+		       SMB_STRUCT_STAT *psbuf)
+{
+	struct smb_filename smb_fname = {
+			.base_name = discard_const_p(char, fname)
+	};
+	int ret;
+
+	if (lp_posix_pathnames()) {
+		ret = SMB_VFS_LSTAT(conn, &smb_fname);
+	} else {
+		ret = SMB_VFS_STAT(conn, &smb_fname);
+	}
+
+	if (ret != -1) {
+		*psbuf = smb_fname.st;
+	}
+	return ret;
+}
+
+/**
  * Ensure LSTAT is called for POSIX paths.
  */
 
